x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1061,4 +1062,155 @@ public static Expression castStringLiteral(Expression from, DataType target) {
             }
         }
     }
+
+    /**
+     * The EsqlIndexResolver will create InvalidMappedField instances for fields that are ambiguous (i.e. have multiple mappings).
+     * During ResolveRefs we do not convert these to UnresolvedAttribute instances, as we want to first determine if they can
+     * instead be handled by conversion functions within the query. This rule looks for matching conversion functions and converts
+     * those fields into MultiTypeEsField, which encapsulates the knowledge of how to convert these into a single type.
+     * This knowledge will be used later in generating the FieldExtractExec with built-in type conversion.
+     * Any fields which could not be resolved by conversion functions will be converted to UnresolvedAttribute instances in a later rule
+     * (See UnresolveUnionTypes below).
+     */
+    private static class ResolveUnionTypes extends BaseAnalyzerRule {
+
+        record TypeResolutionKey(String fieldName, DataType fieldType) {}
+
+        @Override
+        protected LogicalPlan doRule(LogicalPlan plan) {
+            List<FieldAttribute> unionFieldAttributes = new ArrayList<>();
+            // See if the eval function has an unresolved MultiTypeEsField field
+            // Replace the entire convert function with a new FieldAttribute (containing type conversion knowledge)
+            plan = plan.transformExpressionsOnly(
```

Comment:
Approaching in place is short and effective however it has a number of side-effects:
- creates new fields with the same name that clash the original source and need to be managed
- need to be pushed into the underlying source which creates additional input for previous commands. This issue might be hidden due to the fact that the field could not have been used before.
- the analyzer ends up setting up a payload for the fieldExtractor which is somewhat fragile and could be avoided

A better approach would be to extract these conversion fields into a synthetic eval right before its usage and then handle the pushdown of the converted valuesource during physical planning.

For example:
```
from index | where to_ip(client_ip) != "127.0.0.1"
```
becomes
```
from index | eval _ = to_ip(client_ip) | where _ != "127.0.0.1"
```